cpp.c

```diff
@@ -66,11 +51,71 @@ void preprocess_tokens_detail_ifdef_ifndef(const char *keyword)
             char *ident = token->sval;
             if (strcmp("ifdef", ident) == 0 || strcmp("ifndef", ident) == 0)
                 cnt++;
-            else if (strcmp("endif", ident) == 0 && --cnt == 0)
+            else if (strcmp("else", ident) == 0 && --cnt == 0) {
+                if (strcmp("else", keyword) == 0) {
+                    error_unexpected_token_str("#else after #else", token);
+                    break;
+                }
+                else {
+                    preprocess_tokens_detail_else();
+                    break;
+                }
+            }
+            else if (strcmp("endif", ident) == 0 && --cnt == 0) {
+                vector_pop(if_stack);
+                expect_token(tNEWLINE);
                 break;
+            }
         }
     }
+}
+
+void preprocess_tokens_detail_define()
+{
+    char *name = expect_token(tIDENT)->sval;
+    Vector *tokens = new_vector();
+    while (!match_token(tNEWLINE)) vector_push_back(tokens, pop_token());
+    expect_token(tNEWLINE);
+    add_define(name, tokens);
+}
+
+void preprocess_tokens_detail_else()
+{
+    int *tf = vector_peek(if_stack);
+    if (tf == NULL) {
+        error("invalid #else");
+    }
+    if (*tf == 0) {
```

Comment:
Is that safe? `tf` may be 0 or 1, so this line may cause segmentation fault. I wonder why unit tests have passed :thinking:

Reply:
not L85-87, but L88

Reply:
`tf` has `int *` type to detect `NULL`, I wonder.

Reply:
The value of `tf` (not `*tf`) is 0 or 1, right?